### PR TITLE
feat/get available time of facility

### PR DESCRIPTION
**Pull request by new5558**

### CODE REVIEW

File: cmd/helper.go [hunk 1]
```diff
@@ -206,3 +210,75 @@ func isAbleToViewFacilityRequestFull(fs *FacilityServer, userID int64, facilityR
...
+
+	now := time.Now()
+	if helper.DayDifference(now, finishTime) >= 30 {
+		return &typing.InputError{Name: "Booking date can only be within 30 days period from today"}
+	}
+
+	HourStart, _, _ := startTime.Clock()
+
+	dayDifference := helper.DayDifference(now, startTime)
+	if dayDifference < 0 || (dayDifference == 0 && HourStart < now.Hour()) {
```
Comment (new5558) [inline cmd/helper.go]: No need to block booking hour, because we only check for day

File: cmd/helper.go [hunk 2]
```diff
@@ -206,3 +210,75 @@ func isAbleToViewFacilityRequestFull(fs *FacilityServer, userID int64, facilityR
 
 	return handlePermissionChannel(permissionEventChannel, permissionFacilityChannel)
 }
+
+// isAbleToGetAvailableTimeOfFacility a function to check whether user can check facility availability
+func isAbleToGetAvailableTimeOfFacility(startTime time.Time, finishTime time.Time) typing.CustomError {
+	day := helper.DayDifference(startTime, finishTime) + 1
```
Comment (new5558) [inline cmd/helper.go]: Naming of day is not intuitive

File: cmd/helper.go [hunk 3]
```diff
@@ -206,3 +210,75 @@ func isAbleToViewFacilityRequestFull(fs *FacilityServer, userID int64, facilityR
...
+	if dayDifference < 0 || (dayDifference == 0 && HourStart < now.Hour()) {
+		return &typing.InputError{Name: "Booking time must not be in the past"}
+	}
+
+	return nil
+}
+
+// createResultEmptyArray is function to create 2D empy boolean array according to input
+func createResultEmptyArray(startTime time.Time, finishTime time.Time, operatingHours map[int32]*common.OperatingHour) []*facility.GetAvailableTimeOfFacilityResponse_Day {
+	day := helper.DayDifference(startTime, finishTime) + 1
```
Comment (new5558) [inline cmd/helper.go]: Day naming

File: cmd/helper.go [hunk 4]
```diff
@@ -206,3 +210,75 @@ func isAbleToViewFacilityRequestFull(fs *FacilityServer, userID int64, facilityR
...
+		for i, item := range resultArray[index].Items {
+			currentHour := int(startHour) + i
+			if item && currentHour <= requestStartHour || currentHour >= requestFinishHour {
+				resultArray[index].Items[i] = false
+			}
+		}
+
+	}
+
+	return &facility.GetAvailableTimeOfFacilityResponse{Day: resultArray}, nil
```
Comment (new5558) [inline cmd/helper.go]: Why return nil here

File: cmd/main.go [hunk 5]
```diff
@@ -199,6 +200,34 @@ func (fs *FacilityServer) GetFacilityRequestStatusFull(ctx context.Context, in *
 	return result, nil
 }
 
+// GetAvailableTimeOfFacility is a function to get available of facility will ignore hours and seconds in start/finish input
+func (fs *FacilityServer) GetAvailableTimeOfFacility(ctx context.Context, in *facility.GetAvailableTimeOfFacilityRequest) (*facility.GetAvailableTimeOfFacilityResponse, error) {
+	facilityRequests, err := fs.dbs.GetApprovedFacilityRequestList(in.FacilityId, in.Start, in.End)
```
Comment (new5558) [inline cmd/main.go]: Should we do it concurrently?

File: cmd/main.go [hunk 6]
```diff
@@ -199,6 +200,34 @@ func (fs *FacilityServer) GetFacilityRequestStatusFull(ctx context.Context, in *
...
+	startTime, _ := ptypes.Timestamp(in.Start)
+	finishTime, _ := ptypes.Timestamp(in.End)
+	err = isAbleToGetAvailableTimeOfFacility(startTime, finishTime)
+	if err != nil {
+		return nil, status.Error(err.Code(), err.Error())
+	}
+
+	operatingHours := map[int32]*common.OperatingHour{}
+	for _, operatingHour := range facilityInfo.OperatingHours {
+		operatingHours[common.DayOfWeek_value[operatingHour.Day.String()]] = operatingHour
```
Comment (new5558) [inline cmd/main.go]: Can we access enum better than this

File: internal/database/helper.go [hunk 7]
```diff
@@ -95,25 +96,26 @@ func convertFacilityRequestWithInfoModelToProto(data *model.FacilityRequestWithI
...
-	dayStart := start.Day()
-	now := time.Now()
-	if dayStart != finish.Day() {
+	if helper.DayDifference(start, finish) != 0 {
 		return &typing.InputError{Name: "Start and Finish must be the same day"}
 	}
 
-	if dayStart > now.Day()+30 {
+	now := time.Now()
+	dayDifference := helper.DayDifference(now, start)
```
Comment (new5558) [inline internal/database/helper.go]: Naming not intuitive

**Review by boompttn — APPROVED** [review]

**Pull request merged**